text/XXX-flatten-feature.md

```diff
@@ -0,0 +1,325 @@
+- Name: Flatten builders and buildpacks
+- Start Date: 2023-07-13
+- Author(s): @jjbustamante, @dlion
+- Status: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+[summary]: #summary
+
+We propose to add new capabilities to the Pack tool that allow end users to reduce the number of Buildpack's layers in an OCI image by flattening according to their requirements.
+
+# Definitions
+[definitions]: #definitions
+
+- Buildpack: A buildpack is a set of executables that inspects your app source code and creates a plan to build and run your application.
+- Builder: A builder is an image that contains all the components necessary to execute a build. A builder image is created by taking a build image and adding a lifecycle, buildpacks, and files that configure aspects of the build including the buildpack detection order and the location(s) of the run image
+
+# Motivation
+[motivation]: #motivation
+
+- Why should we do this?
+
+There is a limit in the number of layer an image can have, at least on Docker, which is *127*, this feature has being request by the community, issue [#1595](https://github.com/buildpacks/pack/issues/1595), as a workaround to solve error thrown by docker when the limit is reached
+
+- What use cases does it support?
+
+Buildpacks provider like Paketo have composite buildpacks with several layers, when they pull many of those together into a builder, hitting the layer limit for a container image happens very often. A feature for the Buildpack author to group the buildpacks by any attribute will allow them to squash those groups into one layer and reduce their total number of layers, avoiding the layer limit.
+
+- What is the expected outcome?
+
+When Buildpacks Authors execute commands like:
+
+`pack builder create ... <flatten options>` or
+`pack buildpack package ... <flatten options>`
+
+The final OCI image artifact (A) SHOULD contain layers blobs with more than *one* buildpack according to the configuration provided by the user. If we compare an artifact (B) created *without* `<flatten options>` then: 
+
+$numberOfBuildpackLayers(A) \leq numberOfBuildpackLayers(B)$
+
+A and B MUST be otherwise interchangeable, only differing by their number of layers.
+
+
+# What it is
+[what-it-is]: #what-it-is
+
+The proposal is to include new experimental flags to the following commands on pack:
+
+- `pack builder create`
+- `pack buildpack package`
+
+The new flags will move from experimental status to supported status when maintainers deem it appropriate.
+
+The new flags to be included are:
+
+- `--flatten` will flatten all the Buildpacks in one layer
+- `--flatten --depth=<int>` If there are composite buildpack to be flatten, `depth` is the high in the buildpack dependency tree to start flattening the buildpacks
```

Comment:
I would love to know if this is useful in practice. If it is not, could we remove it? As it adds complexity to the code...

Reply:
We are going to talk about it during the Thursday session, let's see how it goes